### ISSUE TITLE
Speech command recogniser: Offline recognition - Input issues

### ISSUE THREAD

**Issue by Arpitrf** [tensorflow/tfjs#3820]
The offline recognition part in  [readme ](https://github.com/tensorflow/tfjs-models/tree/master/speech-commands) mentions that we need to "obtain the spectrogram of an audio snippet through a certain means, e.g., by loading the data from a .wav file or synthesizing the spectrogram programmatically".

Can you please explain how to do that? I am unable to find a way.

Thank you

**Comment by rthadur** [comment]
@Arpitrf did you check example code provided below offline recognition section ?

````
const tf = require('@tensorflow/tfjs');
const speechCommands = require('@tensorflow-models/speech-commands');

const recognizer = speechCommands.create('BROWSER_FFT');

// Inspect the input shape of the recognizer's underlying tf.Model.
console.log(recognizer.modelInputShape());
// You will get something like [null, 43, 232, 1].
// - The first dimension (null) is an undetermined batch dimension.
// - The second dimension (e.g., 43) is the number of audio frames.
// - The third dimension (e.g., 232) is the number of frequency data points in
//   every frame (i.e., column) of the spectrogram
// - The last dimension (e.g., 1) is fixed at 1. This follows the convention of
//   convolutional neural networks in TensorFlow.js and Keras.

// Inspect the sampling frequency and FFT size:
console.log(recognizer.params().sampleRateHz);
console.log(recognizer.params().fftSize);


const x = tf.tensor4d(
    mySpectrogramData, [1].concat(recognizer.modelInputShape().slice(1)));
const output = await recognizer.recognize(x);
// output has the same format as `result` in the online streaming example
// above: the `scores` field contains the probabilities of the words.

tf.dispose([x, output]);
````

**Comment by Arpitrf** [comment]
Yes, but how to obtain **mySpectrogramData** ?

I did the following - 

```
let buffer = fs.readFilSync('zero1.wav');
let input = wav.decode(buffer);

# To make size of input equal to 9976
input = input['channelData'][0].slice(1000, 10976)
const x = tf.tensor(inp, [1].concat(recognizer.modelInputShape().slice(1)));
const output = await recognizer.recognize(x);
```

When using the above (And note that zero1.wav is a file obtained from the training data, so should give high accuracy output), I am getting the following ambiguous output - 
![image](https://user-images.githubusercontent.com/14124093/90591268-51e41180-e200-11ea-9321-0afbbf37f09e.png)


This only means that the input to the recognizer.recognize() is incorrect.

**Comment by Arpitrf** [comment]
Request your help on this!

**Comment by pyu10055** [comment]
@Arpitrf Please checkout https://github.com/tensorflow/tfjs-models/blob/master/speech-commands/training/browser-fft/training_custom_audio_model_in_python.ipynb
There is a saved model named `pre_proc_model`, if you are in node, you can use our saved model direct execution API to process the wav files and use the output of the model as input to the recognizer.

**Comment by Arpitrf** [comment]
I did the following:
```
preproc_model_path = './tfjs-sc-model/sc_preproc_model'
preproc_model = tf.keras.models.load_model(preproc_model_path)


TARGET_SAMPLE_RATE = 44100
# The specific audio tensor length expected by the preprocessing model.
EXPECTED_WAVEFORM_LEN = preproc_model.input_shape[-1]

def read_wav(filepath):
	file_contents = tf.io.read_file(filepath)
	return tf.expand_dims(tf.squeeze(tf.audio.decode_wav(file_contents, desired_channels=-1, desired_samples=TARGET_SAMPLE_RATE).audio, axis=-1), 0)

def crop_and_convert_to_spectrogram(waveform, label):
  cropped = tf.slice(waveform, begin=[0, 0], size=[1, EXPECTED_WAVEFORM_LEN])
  return tf.squeeze(preproc_model(cropped), axis=0), label

audio = read_wav('./zero1.wav')
spec = crop_and_convert_to_spectrogram(audio, 0)
print(spec)
```

**Note: zero1.wav is a file obtained from the actual training dataset that was used by tensorflowJS. (So no issues with the input file)** 
And got the following output: 
```
(<tf.Tensor: shape=(43, 232, 1), dtype=float32, numpy=
array([[[nan],
        [nan],
        [nan],
        ...,
        [nan],
        [nan],
        [nan]],

       [[nan],
        [nan],
        [nan],
        ...,
        [nan],
        [nan],
        [nan]],

       [[nan],
        [nan],
        [nan],
        ...,
        [nan],
        [nan],
        [nan]],

       ...,

       [[nan],
        [nan],
        [nan],
        ...,
        [nan],
        [nan],
        [nan]],

       [[nan],
        [nan],
        [nan],
        ...,
        [nan],
        [nan],
        [nan]],

       [[nan],
        [nan],
        [nan],
        ...,
        [nan],
        [nan],
        [nan]]], dtype=float32)>, 0)
```

Which does not seem right. Request further help.

**Comment by MariamQusai** [comment]
Hi all, 
We need more information on the parameters of the spectrogram to be passed when performing the offline recognition. 
overlap factor, windowing, any smoothing factors.

**Comment by philipclaesson** [comment]
@MariamQusai There's a helpful comment in this [readme](https://github.com/tensorflow/tfjs-models/blob/master/speech-commands/training/browser-fft/README.md):

> The models trained this way expect inputs to be spectrograms in a format consistent with WebAudio's getFloatFrequencyData. Therefore they can be deployed to the browser using the speech-commands library for inference.

and in the notebook
> This preprocessing model replicates WebAudio's AnalyzerNode.getFloatFrequencyData
> (https://developer.mozilla.org/en-US/docs/Web/API/AnalyserNode/getFloatFrequencyData).
> It performs short-time Fourier transform (STFT) using a length-2048 Blackman
> window. It opeartes on mono audio at the 44100-Hz sample rate.

https://en.wikipedia.org/wiki/Window_function#Blackman_window

However, @rthadur do you know if the source code for the preprocessing model available anywhere?

**Comment by MariamQusai** [comment]
@philipclaesson yes I did check the Web audio API [documentation](https://webaudio.github.io/web-audio-api/#current-time-domain-data). 
however, I need to know what parameters have you selected when training the speech-command model, or you went with the defaults. 
Also what is the hop size when performing windowing.

**Comment by Pita** [comment]
I also have issues with the preproc model producing NaNs on certain inputs, even with the example inputs. This is making training very hard. Would like to know as well how the preproc model was produced to find the mistake there

**Comment by Leiga** [comment]
> I did the following:
> 
> ```
> preproc_model_path = './tfjs-sc-model/sc_preproc_model'
> preproc_model = tf.keras.models.load_model(preproc_model_path)
> 
> 
> TARGET_SAMPLE_RATE = 44100
> # The specific audio tensor length expected by the preprocessing model.
> EXPECTED_WAVEFORM_LEN = preproc_model.input_shape[-1]
> 
> def read_wav(filepath):
> 	file_contents = tf.io.read_file(filepath)
> 	return tf.expand_dims(tf.squeeze(tf.audio.decode_wav(file_contents, desired_channels=-1, desired_samples=TARGET_SAMPLE_RATE).audio, axis=-1), 0)
> 
> def crop_and_convert_to_spectrogram(waveform, label):
>   cropped = tf.slice(waveform, begin=[0, 0], size=[1, EXPECTED_WAVEFORM_LEN])
>   return tf.squeeze(preproc_model(cropped), axis=0), label
> 
> audio = read_wav('./zero1.wav')
> spec = crop_and_convert_to_spectrogram(audio, 0)
> print(spec)
> ```
> 
> **Note: zero1.wav is a file obtained from the actual training dataset that was used by tensorflowJS. (So no issues with the input file)**
> And got the following output:
> 
> ```
> (<tf.Tensor: shape=(43, 232, 1), dtype=float32, numpy=
> array([[[nan],
>         [nan],
>         [nan],
>         ...,
>         [nan],
>         [nan],
>         [nan]],
> 
>        [[nan],
>         [nan],
>         [nan],
>         ...,
>         [nan],
>         [nan],
>         [nan]],
> 
>        [[nan],
>         [nan],
>         [nan],
>         ...,
>         [nan],
>         [nan],
>         [nan]],
> 
>        ...,
> 
>        [[nan],
>         [nan],
>         [nan],
>         ...,
>         [nan],
>         [nan],
>         [nan]],
> 
>        [[nan],
>         [nan],
>         [nan],
>         ...,
>         [nan],
>         [nan],
>         [nan]],
> 
>        [[nan],
>         [nan],
>         [nan],
>         ...,
>         [nan],
>         [nan],
>         [nan]]], dtype=float32)>, 0)
> ```
> 
> Which does not seem right. Request further help.

I had the same problem here. Did you solve it?

**Comment by EchoShao8899** [comment]
Can you check if your input tensor of `preproc_model()` contains a lot of zero entries? I met the same problem, and I think it's these zero entries that cause the "nan" problem.

**Comment by Zjq9409** [comment]
> Can you check if your input tensor of `preproc_model()` contains a lot of zero entries? I met the same problem, and I think it's these zero entries that cause the "nan" problem.

it's these zero entries and cause the "nan" problem, so, How solve it?

**Comment by shaoyijia** [comment]
> > Can you check if your input tensor of `preproc_model()` contains a lot of zero entries? I met the same problem, and I think it's these zero entries that cause the "nan" problem.
> 
> it's these zero entries and cause the "nan" problem, so, How solve it?

Hi, I think so. These zero entries will cause FFT to have undesirable behavior. 

You can check our simple fix to avoid this issue (https://github.com/tensorflow/examples/commit/5ba37ab2575827aba7038eca74bee4915e4507af) and see if it can work for your case.
